src/utils.cpp

```diff
@@ -332,8 +332,153 @@ void cnf_to_dimacs(Term cnf, std::ostringstream & y)
+      Term ri_new=give_symbolic_name(ri, true);//The symbolic name of the right child
+      //pushing a pair of c, (a op b) where c is mid, a is le_new and b is ri_new, the new symbolic expressions(or their original symbols if they were originally symbolic constants)
+      reduced.push_back({mid, s->make_term(op, le_new, ri_new)});
+    }
+    else if(op == smt::Not)
+    { //Works in the same way as the if statement above
+      Term t = (*u->begin());
+      Term mid=give_symbolic_name(u, false);
+      Term child=give_symbolic_name(t, true);
+      reduced.push_back({mid, s->make_term(Not, child)});
+    }
+  }
+
+  TermVec clauses;
+  
+  for(auto u:reduced)
+  {
+    Term fi = u.first;
+    Term se = u.second;
+    smt::Op op = se->get_op();
+    if(op == smt::And)
+    {//((~a) v (~b) v (c)) and ((a) v (~c)) and ((b) v (~c))
+      auto it = (se->begin());
+      Term le = (*it);
+      it++;
+      Term ri = (*it);
+      clauses.push_back(s->make_term(Or, s->make_term(Or, s->make_term(Not, le), s->make_term(Not, ri)), fi));
+      clauses.push_back(s->make_term(Or, le, s->make_term(Not, fi)));
+      clauses.push_back(s->make_term(Or, ri, s->make_term(Not, fi)));
+    }
+    else if(op == smt::Or)
+    {//((a) v (b) v (~c)) and ((~a) v c) and ((~b) v c)
+      auto it = (se->begin());
+      Term le = (*it);
+      it++;
+      Term ri = (*it);
+      clauses.push_back(s->make_term(Or, s->make_term(Or, le, ri), s->make_term(Not, fi)));
+      clauses.push_back(s->make_term(Or, s->make_term(Not, le), fi));
+      clauses.push_back(s->make_term(Or, s->make_term(Not, ri), fi));
+    }
+    else if(op == smt::Xor)
+    {//((~a) v (~b) v (~c)) and ((a) v (b) v (~c)) and ((c) v (b) v (~a)) and ((c) v (a) v (~b))
+      auto it = (se->begin());
+      Term le = (*it);
+      it++;
+      Term ri = (*it);
+      clauses.push_back(s->make_term(Or, s->make_term(Or, s->make_term(Not, le), s->make_term(Not, ri)), s->make_term(Not, fi)));
+      clauses.push_back(s->make_term(Or, s->make_term(Or, le, ri), s->make_term(Not, fi)));
+      clauses.push_back(s->make_term(Or, s->make_term(Or, fi, ri), s->make_term(Not, le)));
+      clauses.push_back(s->make_term(Or, s->make_term(Or, fi, le), s->make_term(Not, ri)));
+    }
+    else if(op == smt::Implies)
+    {//((~a) v (b) v (~c)) and ((a) v (c)) and ((~b) v (c))
+      auto it = (se->begin());
+      Term le = (*it);
+      it++;
+      Term ri = (*it);
+      clauses.push_back(s->make_term(Or, s->make_term(Or, s->make_term(Not, le), ri), s->make_term(Not, fi)));
+      clauses.push_back(s->make_term(Or, le, fi));
+      clauses.push_back(s->make_term(Or, s->make_term(Not, ri), fi));
+    }
+    else
+    {//((~a) v (~c)) and ((a) v (c))
+      Term le = (*(se->begin()));
+      clauses.push_back(s->make_term(Or, s->make_term(Not, le), s->make_term(Not, fi)));
+      clauses.push_back(s->make_term(Or, le, fi));  
+    }
+  }
+  //taking the and of all clauses generated to create the cnf
```

Comment:
This is fine, but the following should also work: `return  s->make_term(And, clauses);` If it doesn't, I can fix it :smile:

Reply:
It's working :). actually, I didn't know this was allowed and so now I binarised my formula before doing the tseitin transformation